v2/controller.go

```diff
@@ -0,0 +1,212 @@
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package v2
+
+import (
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	statsv2 "github.com/containerd/cgroups/v2/stats"
+	"github.com/pkg/errors"
+)
+
+type cgValuer interface {
+	Values() []CGStr
+}
+
+type Resources struct {
+	CPU    *CPU
+	Memory *Memory
+	Pids   *Pids
+}
+
+func (r *Resources) Values() (o []CGStr) {
+	values := []cgValuer{
+		r.CPU,
+		r.Memory,
+		r.Pids,
+	}
+	for _, v := range values {
+		if v == nil {
+			continue
+		}
+		o = append(o, v.Values()...)
+	}
+	return o
+}
+
+type CGStr struct {
+	key   string
+	value interface{}
+}
+
+func (c *CGStr) write(path string, perm os.FileMode) error {
+	var data []byte
+	switch t := c.value.(type) {
+	case uint64:
+		data = []byte(strconv.FormatUint(t, 10))
+	case int64:
+		data = []byte(strconv.FormatInt(t, 10))
+	case []byte:
+		data = t
+	case string:
+		data = []byte(t)
+	default:
+		return ErrInvalidFormat
+	}
+	return ioutil.WriteFile(
+		filepath.Join(path, c.key),
+		data,
+		perm,
+	)
+}
+
+func writeCGStr(path string, strs []CGStr) error {
+	for _, o := range strs {
+		if err := o.write(path, defaultFilePerm); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func NewController(mountpoint string, g string, resources *Resources) (*Controller, error) {
+	if err := VerifyGroupPath(g); err != nil {
+		return nil, err
+	}
+
+	path := filepath.Join(mountpoint, string(g))
+	if err := os.MkdirAll(path, defaultDirPerm); err != nil {
+		return nil, err
+	}
+	if err := writeCGStr(path, resources.Values()); err != nil {
+		// clean up cgroup dir on failure
+		os.Remove(path)
+		return nil, err
+	}
+	return &Controller{
+		unifiedMountpoint: mountpoint,
+		path:              path,
+	}, nil
+}
+
+func LoadController(mountpoint string, g string) (*Controller, error) {
+	if err := VerifyGroupPath(g); err != nil {
+		return nil, err
+	}
+	path := filepath.Join(mountpoint, string(g))
+	return &Controller{
+		unifiedMountpoint: mountpoint,
+		path:              path,
+	}, nil
+}
+
+type Controller struct {
+	unifiedMountpoint string
+	path              string
+}
+
+func (c *Controller) NewChild(name string, resources *Resources) (*Controller, error) {
+	if strings.HasPrefix(name, "/") {
+		return nil, errors.New("name must be relative")
+	}
+	path := filepath.Join(c.path, name)
+	if err := os.MkdirAll(path, defaultDirPerm); err != nil {
+		return nil, err
+	}
+	if err := writeCGStr(path, resources.Values()); err != nil {
+		// clean up cgroup dir on failure
+		os.Remove(path)
+		return nil, err
+	}
+	return &Controller{
+		unifiedMountpoint: c.unifiedMountpoint,
+		path:              path,
+	}, nil
+}
+
+func (c *Controller) AddProc(pid uint64) error {
+	v := CGStr{
+		key:   cgroupProcs,
+		value: pid,
+	}
+	return writeCGStr(c.path, []CGStr{v})
+}
+
+func (c *Controller) Delete() error {
+	return remove(c.path)
```

Comment:
can be just `os.Remove`?

Reply:
I think we still need the remove all logic but i'm not sure if this is more robust in v2